tlv/record_type.go

```diff
@@ -0,0 +1,59 @@
+package tlv
+
+import (
+	"github.com/btcsuite/btcd/btcec/v2"
+	"golang.org/x/exp/constraints"
+)
+
+// Primitive is a type constraint that capture the set of "primitive" types,
+// which are the built in stdlib types, and type defs of those types
+type Primitive interface {
+	constraints.Integer | ~[]byte | ~[32]byte | ~[33]byte | ~bool |
+		~*btcec.PublicKey | ~[64]byte
+}
+
+// RecordT is a high-order type makes it easy to encode known "primitive" types
+// as TLV records.
+type RecordT[T Record | Primitive] struct {
+	// Val is the value of the underlying record. Go doesn't let us just
+	// embed the type param as a struct field, so we need an intermediate
+	// variable.
+	Val T
+
+	// Type is the type of the TLV record.
+	Type Type
+}
+
+// NewPrimitiveRecord creates a new RecordT type from a given primitive type.
+func NewPrimitiveRecord[T Primitive](val T, tlvType Type) RecordT[T] {
+	return RecordT[T]{
+		Val:  val,
+		Type: tlvType,
+	}
+}
+
+// NewRecordT creates a new RecordT type from a given Record type. This is
+// useful to wrap a given record in this utility type, which also serves as an
+// extra type annotation. The underlying type of the record is retained.
+func NewRecordT[T Record](record T) RecordT[T] {
+	// Go doesn't yet allow interfaces for union type constraints. So we'll
+	// cast to any, then get a concrete record so we can call the method on
+	// it.
+	tlvRecord := any(record).(Record)
+
+	return RecordT[T]{
+		Val:  record,
+		Type: tlvRecord.Type(),
+	}
+}
+
+// Record returns the underlying record interface for the record type.
+func (t *RecordT[T]) Record() Record {
+	// Go doesn't allow type assertions on a type param, so to work around
+	// this, we'll convert to any, then do our type assertion.
+	if tlvRecord, ok := any(t.Val).(Record); ok {
+		return tlvRecord
+	}
```

Comment:
I think if we go with the altered approach we can omit this exception entirely and the entire body of this function would be "serialize yourself, then prefix with your length, then pull the type annotation from the other field" with no edge cases.